test/gateway/eip20_cogateway/redeem.js

```diff
@@ -0,0 +1,201 @@
+
+const EIP20CoGateway = artifacts.require('EIP20CoGateway'),
+    Core = artifacts.require('MockCore'),
+    MessageBus = artifacts.require('MessageBus'),
+    UtilityToken = artifacts.require('UtilityToken'),
+    EIP20Token = artifacts.require('EIP20Token'),
+    BN = require('bn.js'),
+    MockToken = artifacts.require('MockToken'),
+    Utils = require("../../test_lib/utils"),
+    MockEIP20CoGateway = artifacts.require('MockEIP20CoGateway'),
+    Event = require('../../test_lib/event_decoder');
+
+let eip20CoGateway,
+    stateRoot = "0x70b4172eb30c495bf20b5b12224cd2380fccdd7ffa2292416b9dbdfc8511585d",
+    valueToken,
+    core,
+    organization,
+    gateway,
+    utilityToken,
+    bountyAmount,
+    owner,
+    staker,
+    stakerBalance;
+
+async function _setup(accounts) {
+    
+    valueToken = accounts[0];
+    core = await Core.new(1, 2, 0, stateRoot, accounts[1]);
+    organization = accounts[2];
+    gateway = accounts[3];
+    owner = accounts[8];
+    utilityToken = await MockToken.new({from: owner});
+    bountyAmount = new BN(100);
+    staker = accounts[7];
+    stakerBalance = new BN(1000000);
+
+    eip20CoGateway = await EIP20CoGateway.new(
+        valueToken,
+        utilityToken.address,
+        core.address,
+        bountyAmount,
+        organization,
+        gateway,
+    );
+    
+    await utilityToken.transfer(accounts[7], stakerBalance,{from: owner});
+    
+    await utilityToken.approve(
+        eip20CoGateway.address,
+        new BN(1000),
+        {from: accounts[7]},
+    );
+    
+}
+
+contract('EIP20CoGateway.redeem() ', function (accounts) {
+    
+    let amount,
+        beneficiary = accounts[4],
+        gasPrice = new BN(10),
+        gasLimit = new BN(10),
+        nonce = new BN(1),
+        hashLockObj = Utils.generateHashLock(),
+        hashLock = hashLockObj.l;
+    
+    beforeEach(async function () {
+        
+        await _setup(accounts);
+        amount = new BN(100);
+        
+    });
+    
+    it('should fail when bounty amount is not equal as msg.value', async function () {
+
+        let bounty = new BN(10);
+        await Utils.expectRevert(eip20CoGateway.redeem(
+            amount,
+            beneficiary,
+            gasPrice,
+            gasLimit,
+            nonce,
+            hashLock,
+            { from: accounts[7], value: bounty },
+            ),
+            'msg.value must match the bounty amount',
+        );
+    });
+    
+    it('should fail when amount is 0', async function () {
+
+        amount = 0;
+        await Utils.expectRevert(eip20CoGateway.redeem(
+            amount,
+            beneficiary,
+            gasPrice,
+            gasLimit,
+            nonce,
+            hashLock,
+            { from: accounts[7], value: bountyAmount },
+            ),
+            'Redeem amount must not be zero',
+        );
+    });
+    
+    it('should fail when messagehash is not in undeclared state', async function () {
```

Comment:
Thinking about this test case, maybe this test case is not needed. We can convert this test to a replay attack.